### PR TITLE
Suppress Stalker ARM32 events around exclusive ops

### PR DESCRIPTION
**Pull request by WorksButNotTested**
…tions

### CODE REVIEW

File: gum/backend-arm/gumstalker-arm.c [hunk 1]
```diff
@@ -3624,6 +3694,10 @@ static void
 gum_exec_block_write_arm_exec_event_code (GumExecBlock * block,
                                           GumGeneratorContext * gc)
 {
+  /* Suppress events during exclusive load/store operations */
+  if (gc->exclusive_load_offset != GUM_INSTRUCTION_OFFSET_NONE)
+    return;
```
Comment (oleavr) [inline gum/backend-arm/gumstalker-arm.c:3699]: Would be good to replace this comment with a more descriptive helper function. Perhaps something like:

```C
  if (gum_generator_context_is_timing_sensitive (gc))
    return;
```

**Comment by WorksButNotTested** [comment]
Changes made.

File: gum/backend-arm/gumstalker-arm.c [hunk 2]
```diff
@@ -305,6 +308,9 @@ static gboolean gum_stalker_iterator_arm_next (GumStalkerIterator * self,
     const cs_insn ** insn);
 static gboolean gum_stalker_iterator_thumb_next (GumStalkerIterator * self,
     const cs_insn ** insn);
+static gboolean gum_is_exclusive_store_insn (const cs_insn * insn);
+static void gum_increment_exclusive_load_offset (GumGeneratorContext * gc);
+static gboolean gum_is_exclusive_load_insn (const cs_insn * insn);
```
Comment (oleavr) [inline gum/backend-arm/gumstalker-arm.c:313]: The `load` and `store` helpers should go to the very bottom, after `gum_stalker_is_kuser_helper`. I'd also order them `load` and then `store`.

File: gum/backend-arm/gumstalker-arm.c [hunk 3]
```diff
@@ -305,6 +308,9 @@ static gboolean gum_stalker_iterator_arm_next (GumStalkerIterator * self,
     const cs_insn ** insn);
 static gboolean gum_stalker_iterator_thumb_next (GumStalkerIterator * self,
     const cs_insn ** insn);
+static gboolean gum_is_exclusive_store_insn (const cs_insn * insn);
+static void gum_increment_exclusive_load_offset (GumGeneratorContext * gc);
```
Comment (oleavr) [inline gum/backend-arm/gumstalker-arm.c:312]: This is a method on the `Gum.GeneratorContext`, so the name should reflect that:

`gum_generator_context_advance_exclusive_load_offset`

Methods on that class should also be moved to their own group. (Groups are separated by a blank line up here.) The ordering of the groups should follow the order of the struct typedefs. (Which flow from highest to lowest level.)

File: gum/backend-arm/gumstalker-arm.c [hunk 4]
```diff
@@ -403,6 +409,8 @@ static void gum_exec_block_write_arm_block_event_code (GumExecBlock * block,
     GumGeneratorContext * gc);
 static void gum_exec_block_write_thumb_block_event_code (GumExecBlock * block,
     GumGeneratorContext * gc);
+static gboolean gum_generator_context_is_timing_sensitive (
+    GumGeneratorContext * gc);
```
Comment (oleavr) [inline gum/backend-arm/gumstalker-arm.c:413]: See note above about grouping and ordering.

File: gum/backend-arm/gumstalker-arm.c [hunk 5]
```diff
@@ -1502,17 +1524,73 @@ gum_stalker_iterator_thumb_next (GumStalkerIterator * self,
   return TRUE;
 }
 
+static gboolean
+gum_is_exclusive_store_insn (const cs_insn * insn)
+{
+
```
Comment (oleavr) [inline gum/backend-arm/gumstalker-arm.c:1530]:
```suggestion
```

File: gum/backend-arm/gumstalker-arm.c [hunk 6]
```diff
@@ -1502,17 +1524,73 @@ gum_stalker_iterator_thumb_next (GumStalkerIterator * self,
...
   if (gc->is_thumb)
     gum_stalker_iterator_thumb_keep (self);
   else
     gum_stalker_iterator_arm_keep (self);
 }
 
+static gboolean
+gum_is_exclusive_load_insn (const cs_insn * insn)
+{
+
```
Comment (oleavr) [inline gum/backend-arm/gumstalker-arm.c:1577]:
```suggestion
```

File: gum/backend-arm/gumstalker-arm.c [hunk 7]
```diff
@@ -429,6 +432,15 @@ static void gum_exec_block_thumb_close_prolog (GumExecBlock * block,
...
 
+static void gum_generator_context_advance_exclusive_load_offset (
+    GumGeneratorContext * gc);
+static gboolean gum_generator_context_is_timing_sensitive (
+    GumGeneratorContext * gc);
+
+static gboolean gum_is_exclusive_load_insn (const cs_insn * insn);
+static gboolean gum_is_exclusive_store_insn (const cs_insn * insn);
+
+
```
Comment (oleavr) [inline gum/backend-arm/gumstalker-arm.c:443]:
```suggestion
```

File: gum/backend-arm/gumstalker-arm.c [hunk 8]
```diff
@@ -429,6 +432,15 @@ static void gum_exec_block_thumb_close_prolog (GumExecBlock * block,
 static gboolean gum_stalker_is_thumb (gconstpointer address);
 static gboolean gum_stalker_is_kuser_helper (gconstpointer address);
 
+static void gum_generator_context_advance_exclusive_load_offset (
+    GumGeneratorContext * gc);
+static gboolean gum_generator_context_is_timing_sensitive (
+    GumGeneratorContext * gc);
```
Comment (oleavr) [inline gum/backend-arm/gumstalker-arm.c:438]: These should be before the group above, as they don't belong to any class. (They're free-standing helpers.)

File: gum/backend-arm/gumstalker-arm.c [hunk 9]
```diff
@@ -1506,6 +1532,10 @@ void
 gum_stalker_iterator_keep (GumStalkerIterator * self)
 {
   GumGeneratorContext * gc = self->generator_context;
+  const cs_insn * insn = gc->instruction->ci;
+
+  if (gum_is_exclusive_load_insn (insn))
```
Comment (oleavr) [inline gum/backend-arm/gumstalker-arm.c:1537]: No need for the `insn` variable here as it's only used once.

File: gum/backend-arm/gumstalker-arm.c [hunk 10]
```diff
@@ -3793,3 +3835,59 @@ gum_stalker_is_kuser_helper (gconstpointer address)
...
+    return;
+
+  gc->exclusive_load_offset++;
+
+  if (gc->exclusive_load_offset == 4)
+      gc->exclusive_load_offset = GUM_INSTRUCTION_OFFSET_NONE;
+}
+
+static gboolean
+gum_generator_context_is_timing_sensitive (GumGeneratorContext * gc)
```
Comment (oleavr) [inline gum/backend-arm/gumstalker-arm.c:3852]: I tend to put getters before regular methods, so this one should go above `advance_exclusive_load_offset()`.

File: gum/backend-arm/gumstalker-arm.c [hunk 11]
```diff
@@ -3793,3 +3835,59 @@ gum_stalker_is_kuser_helper (gconstpointer address)
...
+{
+  return (gc->exclusive_load_offset != GUM_INSTRUCTION_OFFSET_NONE);
```
Comment (oleavr) [inline gum/backend-arm/gumstalker-arm.c:3854]:
```suggestion
  return gc->exclusive_load_offset != GUM_INSTRUCTION_OFFSET_NONE;
```

File: gum/backend-arm/gumstalker-arm.c [hunk 12]
```diff
@@ -3769,6 +3809,24 @@ gum_exec_block_thumb_close_prolog (GumExecBlock * block,
...
+static void
+gum_generator_context_advance_exclusive_load_offset (GumGeneratorContext * gc)
+{
+  if (gc->exclusive_load_offset == GUM_INSTRUCTION_OFFSET_NONE)
+    return;
+
+  gc->exclusive_load_offset++;
+
+  if (gc->exclusive_load_offset == 4)
+      gc->exclusive_load_offset = GUM_INSTRUCTION_OFFSET_NONE;
```
Comment (oleavr) [inline gum/backend-arm/gumstalker-arm.c:3827]:
```suggestion
    gc->exclusive_load_offset = GUM_INSTRUCTION_OFFSET_NONE;
```

**Pull request merged**